File: python/src/deepl_fix.py
import deepl
from dotenv import load_dotenv
import os
from typing import Literal

load_dotenv()

auth_key = os.getenv("DEEPL_API_KEY")
deepl_client = deepl.DeepLClient(auth_key)


def fix_text(text: str, targetLang: Literal["DE", "EN-US"] = "DE") -> str:
    # using back and forth translation, becuase rephrase is only available in DeepL Pro
    langs = ["EN-US", "DE"]
    # ensure targetLang is the last element in the langs array
    langs.remove(targetLang)
    langs.append(targetLang)

    for lang in langs:
        trans = deepl_client.translate_text(text, target_lang=lang)
        text = trans.text

    return text

File: electron-app/src/deepl-fix.ts
import { Translator, TargetLanguageCode } from "deepl-node";

export type FixTextFn = (
  text: string,
  targetLang?: TargetLanguageCode
) => Promise<string>;

export const fixTextFactory = (apiKey: string) => {
  const translator = new Translator(apiKey);

  const fixText: FixTextFn = async (text, targetLang = "de") => {
    const langs = (["en-US", "de"] as TargetLanguageCode[]).filter(
      (lang) => lang !== targetLang
    );
    langs.push(targetLang);

    let fixedText = text;
    for (const lang of langs) {
      const result = await translator.translateText(fixedText, null, lang);
      fixedText = result.text;
    }

    return fixedText;
  };

  return fixText;
};
